src/Workspaces/CSharp/Portable/Extensions/DefaultExpressionSyntaxExtensions.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Utilities;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.CSharp.Extensions
+{
+    internal static class DefaultExpressionSyntaxExtensions
+    {
+        private static readonly LiteralExpressionSyntax s_defaultLiteralExpression = SyntaxFactory.LiteralExpression(SyntaxKind.DefaultLiteralExpression);
+
+        public static bool CanReplaceWithDefaultLiteral(
+            this DefaultExpressionSyntax defaultExpression,
+            CSharpParseOptions parseOptions,
+            OptionSet options,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            if (parseOptions.LanguageVersion >= LanguageVersion.CSharp7_1 &&
+                options.GetOption(CSharpCodeStyleOptions.PreferDefaultLiteral).Value)
+            {
+                var speculationAnalyzer = new SpeculationAnalyzer(
+                    defaultExpression, s_defaultLiteralExpression, semanticModel,
+                    cancellationToken,
+                    skipVerificationForReplacedNode: true,
+                    failOnOverloadResolutionFailuresInOriginalCode: true);
+
+                return !speculationAnalyzer.ReplacementChangesSemantics();
```

Comment:
❓ Can we avoid the use of `SpeculationAnalyzer` for the following cases?

* Assignment of `default(T)` to a new variable in a variable initializer for either of these cases (cannot be a default literal expression):
  * `var` is used or the static type
  * The static type of the variable is not `T`
* Use of `default(T)` as the default value for a parameter (default literal expression allowed when `T` is a reference type or `T` is the type of the parameter)
* Initialization of a field to `default(T)` (same rules as parameters)

:memo: `SpeculationAnalyzer` is *extremely* costly. If we can avoid using it for a case where we know it will either succeed or fail, it is important to do so. For example, see #19258.